compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/BindsMethodValidator.kt

```diff
@@ -54,70 +79,71 @@ internal class BindsMethodValidator : PrivateCodeGenerator() {
-      val paramSuperTypes = (function.parameterSuperTypes() ?: function.receiverSuperTypes())!!
-        .map { it.shortName }
-        .toList()
+    val parameterSuperTypes = function.singleParameterTypeOrNull().superTypes()
+    val receiverSuperTypes = function.receiverType().superTypes()
 
-      val superTypesMessage = if (paramSuperTypes.size == 1) {
-        "has no supertypes."
-      } else {
-        "only has the following supertypes: ${paramSuperTypes.drop(1)}"
-      }
-      throw AnvilCompilationExceptionFunctionReference(
-        message = "@Binds methods' parameter type must be assignable to the return type. " +
-          "Expected binding of type $returnType but impl parameter of type " +
-          "${paramSuperTypes.first()} $superTypesMessage",
-        functionReference = function,
-      )
+    when (returnTypeName) {
+      in parameterSuperTypes -> return
+      in receiverSuperTypes -> return
+      else -> Unit
     }
-  }
 
-  private fun MemberFunctionReference.Psi.parameterMatchesReturnType(): Boolean {
-    return parameterSuperTypes()
-      ?.contains(returnType().asClassReference())
-      ?: false
+    val paramSuperTypes =
+      parameterSuperTypes
+        .ifEmpty { receiverSuperTypes }
+        .map { it.toStringOfSimpleNamesOnly() }
+        .toList()
+
+    throw AnvilCompilationExceptionFunctionReference(
+      message = Errors.bindsParameterMustBeAssignable(
+        paramSuperTypeNames = paramSuperTypes.drop(1),
+        returnTypeName = returnTypeName.toStringOfSimpleNamesOnly(),
+        parameterType = paramSuperTypes.first(),
+      ),
+      functionReference = function,
+    )
   }
 
-  private fun MemberFunctionReference.Psi.parameterSuperTypes(): Sequence<ClassReference>? {
-    return parameters.singleOrNull()
-      ?.type()
-      ?.asClassReference()
-      ?.allSuperTypeClassReferences(includeSelf = true)
+  private fun MemberFunctionReference.Psi.singleParameterTypeOrNull(): TypeReference? {
+    return parameters.singleOrNull()?.type()
   }
 
-  private fun MemberFunctionReference.Psi.receiverMatchesReturnType(): Boolean {
-    return receiverSuperTypes()
-      ?.contains(returnType().asClassReference())
-      ?: false
+  private fun MemberFunctionReference.Psi.receiverType(): TypeReference? {
+    return function.receiverTypeReference?.toTypeReference(declaringClass, module)
   }
 
-  private fun MemberFunctionReference.Psi.receiverSuperTypes(): Sequence<ClassReference>? {
-    return function.receiverTypeReference
-      ?.toTypeReference(declaringClass, module)
-      ?.asClassReference()
-      ?.allSuperTypeClassReferences(includeSelf = true)
+  private fun TypeReference?.superTypes(): Sequence<TypeName> {
+    this ?: return emptySequence()
+    return sequence {
+      yield(this@superTypes)
+      yieldAll(
+        this@superTypes
+          .asClassReference()
+          .directSuperTypeReferences(),
+      )
+    }.map { it.asTypeName() }
```

Comment:
I might be missing something, but it seems like we should continue to use `allSuperTypeClassReferences(includeSelf = true)`.  This version only returns the type and its immediate supertypes, whereas `allSuperTypeClassReferences()` is recursive.

Reply:
Unfortunately, `allSuperTypeClassReferences` lose generic type information (I assume that's because `ClassReference` can't contain (or should not contain?) information about which types are used as type arguments in specific binding declaration. That's why I'm using `TypeReference`s here instead of `ClassReference`s
I will try to make a similar to `allSuperTypeClassReferences` function which returns `TypeReference`s instead of `ClassReference`s

Reply:
Nevermind, seems that I was wrong

Reply:
Also, I'm not quite sure that `allSuperTypeClassReferences` is recursive. Or, at least, I don't see it 🙂